k9mail-library/src/main/java/com/fsck/k9/mail/transport/SmtpTransport.java

```diff
@@ -741,25 +690,42 @@ private CommandResponse executeSimpleCommandWithResponse(String command, boolean
         }
 
         int replyCode = -1;
-        String message = line;
         if (length >= 3) {
             try {
                 replyCode = Integer.parseInt(line.substring(0, 3));
             } catch (NumberFormatException e) { /* ignore */ }
+        }
 
-            if (length > 4) {
-                message = line.substring(4);
+        char c = line.charAt(0);
+        if ((c == '4') || (c == '5')) {
+            if (mEnhancedStatusCodesProvided) {
+                throw buildEnhancedNegativeSmtpReplyException(replyCode, results);
             } else {
-                message = "";
+                throw new NegativeSmtpReplyException(replyCode, join(results, " "));
             }
         }
 
-        char c = line.charAt(0);
-        if ((c == '4') || (c == '5')) {
-            throw new NegativeSmtpReplyException(replyCode, message);
+        return new CommandResponse(replyCode, results);
+    }
+
+    private MessagingException buildEnhancedNegativeSmtpReplyException(int replyCode, List<String> results) {
+        SmtpEnhancedStatusCodeClass escClass = null;
+        SmtpEnhancedStatusCodeSubject escSubject = null;
+        SmtpEnhancedStatusCodeDetail escDetail = null;
+
+        String message = "";
+        for (String resultLine: results) {
+            message += resultLine.split(" ", 2)[1] + " ";
+        }
+        if (results.size() > 0) {
+            String[] esc = results.get(0).split(" ", 2)[0].split("\\.");
+
+            escClass = SmtpEnhancedStatusCodeClass.parse(esc[0]);
+            escSubject = SmtpEnhancedStatusCodeSubject.parse(esc[1]);
+            escDetail = SmtpEnhancedStatusCodeDetail.parse(escSubject, esc[2]);
         }
 
-        return new CommandResponse(replyCode, message);
+        return new EnhancedNegativeSmtpReplyException(replyCode, escClass, escSubject, escDetail, message.trim());
```

Comment:
might as well put the `message.trim()` up there with the other message formatting things